audio/OpenAL/OpenALAudioProvider.cpp

```diff
@@ -0,0 +1,245 @@
+// Copyright © Matt Jones and Contributors. Licensed under the MIT Licence (MIT). See LICENCE.md in the repository root
+// for more information.
+#include <AL/alext.h>
+#include <NovelRT/Audio/OpenAL/OpenALAudioProvider.hpp>
+#include <NovelRT/Exceptions/Exceptions.h>
+#include <spdlog/sinks/stdout_color_sinks.h>
+
+namespace NovelRT::Audio::OpenAL
+{
+    typedef void (ALC_APIENTRY*LoggingCallback)(void*, char, const char*, int) noexcept;
+    typedef void (ALC_APIENTRY*CallbackProvider)(LoggingCallback, void*) noexcept;
+
+    OpenALAudioProvider::OpenALAudioProvider():
+    _sources(std::vector<uint32_t>()),
+    _buffers(std::vector<uint32_t>()),
+    _logger(spdlog::stdout_color_mt("OpenAL"))
+    {
+        //Logger init
+        _logger->set_level(spdlog::level::debug);
+        CallbackProvider setLogCallback = reinterpret_cast<CallbackProvider>(alcGetProcAddress(nullptr, "alsoft_set_log_callback"));
+        if (setLogCallback != nullptr)
+        {
+            setLogCallback(static_cast<LoggingCallback>([](void* ptr, char l, const char* m, int) noexcept
+            {
+                auto provider = reinterpret_cast<OpenALAudioProvider*>(ptr);
+                provider->LogOpenALMessages(l, m);
+            }), this);
+        }
+
+
+        //Device and Context Init
+        _device = alcOpenDevice(nullptr);
+        if(!_device)
+        {
+            std::string error = GetALError();
+            throw Exceptions::InitialisationFailureException(
+                "OpenAL failed to create an audio device!", error);
+        }
+        _context = alcCreateContext(_device, nullptr);
+        if(!_context)
+        {
+            std::string error = GetALError();
+            throw Exceptions::InitialisationFailureException(
+                "OpenAL failed to create and attach a proper context!", error);
+        }
+        alcMakeContextCurrent(_context);
+    }
+
+    OpenALAudioProvider::~OpenALAudioProvider()
+    {
+        Dispose();
+    }
+
+    void OpenALAudioProvider::Dispose()
+    {
+        alGetError();
+        alSourceStopv(static_cast<int>(_sources.size()), reinterpret_cast<ALuint*>(_sources.data()));
+        GetALError();
```

Comment:
Can I ask what the difference is between `alGetError` and `GetALError`?

Reply:
One gets the error code from OpenAL and clears the error register internally (from OpenAL's side) - the other (`GetALError`) does that plus actually parses the error code into something human-readable